text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+
+Unlike most locking mechanisms, a sequence lock doesn't prevent a race
+to access the data it projects.
+Instead, it detects a race only after the load operation already happened,
+and retries it if the load operation raced with a write operation.
+
+A sequence lock in Rust looks something like this:
+
+```rust
+// Incomplete example
+
+pub struct SeqLock<T> {
+    seq: AtomicUsize,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl Sync<T: Copy + Send> for SeqLock<T> {}
+
+impl<T: Copy> SeqLock<T> {
+    /// Safety: Only call from one thread.
+    pub unsafe fn write(&self, value: T) {
+        self.seq.fetch_add(1, Relaxed);
+        write_data(&mut self.data, value, Release);
+        self.seq.fetch_add(1, Release);
+    }
+
+    pub fn read(&self) -> T {
+        loop {
+            let s1 = self.seq.load(Acquire);
+            let data = read_data(&self.data, Acquire);
+            let s2 = self.seq.load(Relaxed);
+            if s1 & 1 == 0 && s1 == s2 {
+                return unsafe { assume_valid(data) };
+            }
+        }
+    }
+}
+```
+
+The `write_data` and `read_data` calls can happen concurrently.
+The `write` method increments the counter before and after,
+such that the counter is odd during `write_data`.
+The `read` function will repeat `read_data` until
+the counter was identical and even both before and after reading.
+This way, `assume_valid` is only ever called on data that was
+not the result of a race.
+
+A big question is how to implement `write_data`, `read_data`, and `assume_valid`
+in Rust in an efficient way while satisfying the memory model.
+
+The somewhat popular `seqlock` crate and similar implementations found in the ecosystem
+all use a regular non-atomic write (preceded by an atomic fence) for writing,
+and `ptr::read_volatile` (followed by an atomic fence) for reading.
+This "works" "fine", but is technically undefined behavior.
+
+The C++ and Rust memory model don't allow for data races,
+so don't allow for a data race to be detected after the fact;
+that's too late.
+
+All of the data would have to be written and read through
+atomic operations to prevent such a data race.
+We don't need the atomicity of the data as a whole though;
+it's fine if there's tearing, since we re-start on a race anyway.
+
+# The C++ Solution
+
+C++'s [P1478] proposes the addition of these two functions to the C++ standard
+library to solve this problem:
+
+```cpp
+void *atomic_load_per_byte_memcpy(void *dest, const void *source, size_t, memory_order);
+void *atomic_store_per_byte_memcpy(void *dest, const void *source, size_t, memory_order);
+```
+
+The first one is effectively a series of `AtomicU8::load`s followed by a memory fence,
+while the second one is basically a memory fence followed by series of `AtomicU8::store`s.
+Except the implementation can be much more efficient.
+The implementation is allowed to load/store the bytes in any order,
+and doesn't have to operate on individual bytes.
+
```

Comment:
I think mixed accesses have to be forbidden still. Even if we want to support them in a future memory model (unclear), I believe the C++ spec for this was pretty careful to forbid them.

Currently allowing them will also break things like tsan, if nothing else... and the fact that they're also more or less forbidden by even Intel's arch manual makes allowing them in the memory model hard to argue for.

Anyway, the current API doesn't allow anything like this, but I suppose I agree this should be documented as intentional.

Reply:
imho mixed atomics should be fine...wasm and javascript have them after all, even on intel.

imho C++ doesn't have them because they didn't want to go through the hassle of specifying how they'd work and C++ has TBAA, making mixing between two non-char[^1] types illegal even with normal load/stores on a single thread.

[^1]: and unsigned char and std::byte_t iirc

Reply:
Mixed atomics also don't work with locks, right? C++ allows locks for it's atomic implementation.

Reply:
> imho mixed atomics should be fine...

From the x86 architecture manual 3A/I 8.1.2.2:

> Software should access semaphores (shared memory used for signalling between multiple processors) using identical addresses and operand lengths. For example, if one processor uses accesses a semaphore using a word access, other processors should not access the semaphore using a byte access.

It does work in practice though, for the most part. I'm not sure we want to rely on this given that the architecture manual says not to do it.

Reply:
I agree for now we should keep ruling them out, because I assume that's what C++ does -- and that should probably be explicitly documented for atomic memcpy, because that operation is much more likely to be used in unsafe type-punning code than the regular Atomic types.

> imho mixed atomics should be fine...wasm and javascript have them after all, even on intel.

Would be interesting to see what wasm/JS engines do with them on intel, if they just ignore what intel says in their manual or if they took some precautions. But that is probably a separate discussion.

Reply:
> AFAIK C++ rules out such mixed accesses

The proposed atomic memcpy in C++ takes `void*`, so typed memory is of no help here. The x86 restriction might have slipped through their meetings, otherwise the proposal would've at least mentioned it. I've started a discussion on their mailing list https://lists.isocpp.org/std-proposals/2023/04/6235.php. If they can convince Intel to drop that line, good for us too.

Reply:
> It does work in practice though, for the most part. I'm not sure we want to rely on this given that the architecture manual says not to do it.

Apparently this is _mostly_ reliable: https://github.com/rust-lang/unsafe-code-guidelines/issues/345